module/zfs/vdev_raidz_math.c

```diff
@@ -0,0 +1,626 @@
+	 * to run "original" implementation (bench_rm->rm_ops = NULL)
+	 */
+	for (impl = 0; impl <= raidz_supp_maths_cnt; impl++) {
+		/* set an implementation to benchmark */
+		curr_impl = raidz_supp_maths[impl];
+		bench_rm->rm_ops = curr_impl;
+
+		run_cnt = 0;
+		t_start = gethrtime();
+
+		do {
+			for (i = 0; i < 50; i++, run_cnt++)
+				bench_fn(bench_rm, fn);
+
+			t_diff = gethrtime() - t_start;
+		} while (t_diff < BENCH_NS);
+
+		speed = run_cnt * BENCH_ZIO_SIZE * NANOSEC;
+		speed /= (t_diff * BENCH_COLS);
+
+		if (bench_fn == __benchmark_gen)
+			raidz_impl_kstats[impl].gen[fn].value.ui64 = speed;
+		else
+			raidz_impl_kstats[impl].rec[fn].value.ui64 = speed;
+
+		/* if curr_impl==NULL the original impl is benchmarked */
+		if (curr_impl != NULL && speed > best_speed) {
+			best_speed = speed;
+
+			if (bench_fn == __benchmark_gen)
+				vdev_raidz_fastest_impl.gen[fn] =
+				    curr_impl->gen[fn];
+			else
+				vdev_raidz_fastest_impl.rec[fn] =
+				    curr_impl->rec[fn];
+		}
+	}
+}
+
+void
+vdev_raidz_math_init(void)
+{
+	raidz_math_ops_t *curr_impl;
+	zio_t *bench_zio = NULL;
+	raidz_map_t *bench_rm = NULL;
+	uint64_t bench_parity;
+	int i, c, fn;
+
+	/* init & vdev_raidz_impl_lock */
+	rw_init(&vdev_raidz_impl_lock, NULL, RW_DEFAULT, NULL);
+
+	/* move supported impl into raidz_supp_maths */
+	for (i = 0, c = 0; i < ARRAY_SIZE(raidz_all_maths); i++) {
+		curr_impl = (raidz_math_ops_t *) raidz_all_maths[i];
+
+		/* initialize impl */
+		if (curr_impl->init)
+			curr_impl->init();
+
+		if (curr_impl->is_supported()) {
+			/* init kstat */
+			init_raidz_kstat(&raidz_impl_kstats[c],
+			    curr_impl->name);
+			raidz_supp_maths[c++] = (raidz_math_ops_t *) curr_impl;
+		}
+	}
+	raidz_supp_maths_cnt = c;	/* number of supported impl */
+	raidz_supp_maths[c] = NULL;	/* sentinel */
+
+#if !defined(_KERNEL)
+	/*
+	 * It is possible that this routine is called multiple times
+	 * inside user-space test. Check if we already initialized the
+	 * fastest ops and skip the benchmark.
+	 */
+	if (raidz_math_initialized) {
+		goto init_done;
+	}
+#endif
+
+	/* init kstat for original routines */
+	init_raidz_kstat(&(raidz_impl_kstats[raidz_supp_maths_cnt]),
+	    "original");
+
+	/* Fake an zio and run the benchmark on it */
+	bench_zio = kmem_zalloc(sizeof (zio_t), KM_SLEEP);
+	bench_zio->io_offset = 0;
+	bench_zio->io_size = BENCH_ZIO_SIZE; /* only data columns */
+	bench_zio->io_data = zio_data_buf_alloc(BENCH_ZIO_SIZE);
+	VERIFY(bench_zio->io_data);
+
+	/* Benchmark parity generation methods */
+	for (fn = 0; fn < RAIDZ_GEN_NUM; fn++) {
+		bench_parity = fn + 1;
+		/* New raidz_map is needed for each generate_p/q/r */
+		bench_rm = vdev_raidz_map_alloc(bench_zio, 9,
+		    BENCH_D_COLS + bench_parity, bench_parity);
+
+		_benchmark_raidz(bench_rm, fn, __benchmark_gen);
+
+		vdev_raidz_map_free(bench_rm);
+	}
+
+	/* Benchmark data reconstruction methods */
+	bench_rm = vdev_raidz_map_alloc(bench_zio, 9, BENCH_COLS, PARITY_PQR);
```

Comment:
Thanks for cleaning this up!  The benchmarking code is much better after you refactored it!